Source/JavaScriptCore/builtins/ModuleLoader.js

```diff
@@ -225,8 +225,137 @@ function requestInstantiate(entry, parameters, fetcher)
+    // satisfy in spec see:
+    //  1. https://tc39.es/ecma262/#sec-import-calls
+    //  2. https://tc39.es/ecma262/#sec-ContinueDynamicImport
+    //  3. https://tc39.es/ecma262/#sec-LoadRequestedModules
+    //  4. https://tc39.es/ecma262/#sec-InnerModuleLoading
+    //
+    // [1] Counter Example
+    //
+    // Given module dependency graph:
+    //
+    //   r1 ---> b ---> c
+    //    |      ^
+    //    |      |
+    //    -----> a <--- r2
+    // 
+    // Note that here we treat requestInstantiate as a background execution for simplification. 
+    // And requestInstantiate1 is the 1st requestInstantiate call in requestSatisfyUtil and 
+    // requestInstantiate2 is the 2nd requestInstantiate call for visited depEntry.
+    // `->` means goto in below steps.
+    // 
+    // Step 1: Dynamically import r1 and r2
+    //         -> requestImportModule(r1) -> requestSatisfyUtil(r1, v1) -> v1.add(r1) -> requestInstantiate1(r1)
+    //         -> requestImportModule(r2) -> requestSatisfyUtil(r2, v2) -> v2.add(r2) -> requestInstantiate1(r2)
+    //     Background Executions = [ requestInstantiate1(r1), requestInstantiate1(r2) ]
+    //
+    // Step 2: requestInstantiate1(r1) is done then go for it's dependencies [ b, a ]
+    //         -> !v1.has(b) -> requestSatisfyUtil(b, v1) -> v1.add(b) -> requestInstantiate1(b)
+    //         -> !v1.has(a) -> requestSatisfyUtil(a, v1) -> v1.add(a) -> requestInstantiate1(a)
+    //     Background Executions = [ requestInstantiate1(r2), requestInstantiate1(b), requestInstantiate1(a) ]
+    //
+    // Step 3: requestInstantiate1(b) is done then go for it's dependencies [ c ]
+    //         -> !v1.has(c) -> requestSatisfyUtil(c, v1) -> v1.add(c) -> requestInstantiate1(c)
+    //     Background Executions = [ requestInstantiate1(r2), requestInstantiate1(a), requestInstantiate1(c) ]
+    //
+    // Step 4: requestInstantiate1(a) is done then go for it's dependencies [ b ]
+    //         -> v1.has(b) -> requestInstantiate2(b)
+    //         -> Since requestInstantiate1(b) cached in Step 3 and b is the only dependency a has, requestSatisfyUtil(a, v1) is fulfilled and cached to Entry(a).satisfy.
+    //     Background Executions = [ requestInstantiate1(r2), requestInstantiate1(c) ]
+    //
+    // Step 5: requestInstantiate1(r2) is done then go for it's dependencies [ a ]
+    //         -> !v2.has(a) -> requestSatisfyUtil(a, v2)
+    //         -> Since requestSatisfyUtil(a) is cached in Step 4 and a is the only dependency r2 has, requestSatisfyUtil(r2, v2) is fulfilled and cached.
+    //         -> linkAndEvaluateModule(r2) -> link(r2) crashes since c is not instantiated.
+    //     Background Executions = [ requestInstantiate1(c) ]
+    // 
+
+    "use strict";
+    
+    var satisfyingEntries = new @Set;
+    return this.requestSatisfyUtil(entry, parameters, fetcher, visited, satisfyingEntries).then((entry) => {
+        for (var satisfyingEntry of satisfyingEntries) {
```

Comment:
We need to use `@forEach` to avoid user-observable side effect. Users can override `Symbol.iterator` method here so that user can collapse module loader.